marlowe-runtime/load-api/Language/Marlowe/Protocol/Load/Client.hs

```diff
@@ -0,0 +1,302 @@
+  where
+  peerProcessing :: SNode node -> ClientStProcessing node m a -> PeerTraced MarloweLoad 'AsClient ('StProcessing node) r m a
+  peerProcessing node ClientStProcessing{..} = AwaitTraced (ServerAgency $ TokProcessing node) \case
+    MsgResume (Succ n) -> Receive $ EffectTraced $ peerCanPush n node <$> recvMsgResume (Succ n)
+
+  peerCanPush :: Nat n -> SNode node -> ClientStCanPush n node m a -> PeerTraced MarloweLoad 'AsClient ('StCanPush n node) r m a
+  peerCanPush n node = \case
+    PushClose next ->
+      YieldTraced tok MsgPushClose $ Cast $ peerPop n node next
+    PushPay payor payee token value next ->
+      YieldTraced tok (MsgPushPay payor payee token value) $ Cast $ peerPush n (SPayNode node) next
+    PushIf cond next ->
+      YieldTraced tok (MsgPushIf cond) $ Cast $ peerPush n (SIfLNode node) next
+    PushWhen timeout next ->
+      YieldTraced tok (MsgPushWhen timeout) $ Cast $ peerPush n (SWhenNode node) next
+    PushCase action next -> case node of
+      SWhenNode node' -> EffectTraced $ YieldTraced tok (MsgPushCase action) . Cast . peerPush n (SCaseNode node') <$> next
+    PushLet valueId value next ->
+      YieldTraced tok (MsgPushLet valueId value) $ Cast $ peerPush n (SLetNode node) next
+    PushAssert obs next ->
+      YieldTraced tok (MsgPushAssert obs) $ Cast $ peerPush n (SAssertNode node) next
+    where
+      tok = ClientAgency $ TokCanPush n node
+
+  peerPop
+    :: Nat n
+    -> SNode node
+    -> ClientStPop n node m a
+    -> PeerTraced MarloweLoad 'AsClient (Pop n node) r m a
+  peerPop n node client = case node of
+    SRootNode -> peerComplete client
+    SPayNode node' -> peerPop n node' client
+    SIfLNode node' -> peerPush n (SIfRNode node') client
+    SIfRNode node' -> peerPop n node' client
+    SWhenNode node' -> peerPop n node' client
+    SCaseNode node' -> peerPush n (SWhenNode node') client
+    SLetNode node' -> peerPop n node' client
+    SAssertNode node' -> peerPop n node' client
+
+  peerPush
+    :: Nat n
+    -> SNode node
+    -> ClientStPush n node m a
+    -> PeerTraced MarloweLoad 'AsClient (Push n node) r m a
+  peerPush n node client = case n of
+    Zero -> peerProcessing node client
+    Succ n' -> peerCanPush n' node client
+
+  peerComplete :: ClientStComplete m a -> PeerTraced MarloweLoad 'AsClient 'StComplete r m a
+  peerComplete ClientStComplete{..} = AwaitTraced (ServerAgency TokComplete) \case
+    MsgComplete hash -> Closed TokDone $ recvMsgComplete hash
+
+-- | Load a contract into the runtime in a space-efficient way. Note: this
+-- relies on the input contract being lazily evaluated and garbage collected as
+-- it is processed. This means that the caller of this function should avoid holding
+-- onto a reference to the input contract (e.g. in a let-binding). Preferably,
+-- the return value of a function that generates the contract should be passed
+-- directly to this function.
+pushContract
```

Comment:
I think that the protocol should include a form of `pushContract` that merkleizes each sub-contract within the contract and returns `Map DatumHash Contract`. This supports the second of the two use cases that we have for merkleization:
1. The client incrementally builds the merkleized contract through repeated pushes as they are building it.
2. The client creates a fully unmerkleized contract but needs the protocol to deeply merkelize it so it fits on the chain.

The first use case occurs when, from necessity, the contract is too large to ever represent (in memory or on disk) as unmerkleized. The second use case occurs when the contract is small enough to create an manipulate in memory, but it is too big to fit in a datum.

Reply:
This is a great idea, as it would be far more performant for contracts that can fit in memory.